src/scf/self_consistent_field.jl

```diff
@@ -69,73 +71,98 @@ Solve the Kohn-Sham equations with a SCF algorithm, starting at ρ.
+
         n_iter += 1
-        ρin = from_real(basis, x)
+        if n_spin == 2  # TODO So ugly
+            # x has 2 blocks: total and spin density
+            ρin, ρspinin = from_real(basis,x[:, :, :, 1]), from_real(basis,x[:, :, :, 2])
+        else
+            ρin, ρspinin = from_real(basis, x), nothing
+        end
 
         # Build next Hamiltonian, diagonalize it, get ρout
         if n_iter == 1 # first iteration
             _, ham = energy_hamiltonian(basis, nothing, nothing;
-                                        ρ=ρin, eigenvalues=nothing, εF=nothing)
+                                        ρ=ρin, ρspin=ρspinin, eigenvalues=nothing, εF=nothing)
         else
             # Note that ρin is not the density of ψ, and the eigenvalues
             # are not the self-consistent ones, which makes this energy non-variational
             energies, ham = energy_hamiltonian(basis, ψ, occupation;
-                                               ρ=ρin, eigenvalues=eigenvalues, εF=εF)
+                                               ρ=ρin, ρspin=ρspinin, eigenvalues=eigenvalues, εF=εF)
         end
 
         # Diagonalize `ham` to get the new state
         nextstate = next_density(ham; n_bands=n_bands, ψ=ψ, eigensolver=eigensolver,
                                  miniter=1, tol=determine_diagtol(info),
                                  n_ep_extra=n_ep_extra,
                                  occupation_function=occupation_function)
-        ψ, eigenvalues, occupation, εF, ρout = nextstate
-        enforce_symmetry && (ρout = DFTK.symmetrize(ρout))
+        ψ, eigenvalues, occupation, εF, ρout, ρspinout = nextstate
+
+        # Update info with results gathered so far
+        info = (ham=ham, basis=basis, converged=converged, stage=:iterate,
+                ρin=ρin, ρspinin=ρspinin, n_iter=n_iter, nextstate...)
+
+        if enforce_symmetry
+            @assert model.spin_polarization in (:none, :spinless)
+            info = merge(info, (ρout=DFTK.symmetrize(ρout), ))
+        end
 
         # Compute the energy of the new state
         if compute_consistent_energies
             energies, _ = energy_hamiltonian(basis, ψ, occupation;
-                                             ρ=ρout, eigenvalues=eigenvalues, εF=εF)
+                                             ρ=ρout, ρspin=ρspinout, eigenvalues=eigenvalues, εF=εF)
         end
-
-        # Update info with results gathered so far
-        info = (ham=ham, basis=basis, energies=energies, converged=false, ρin=ρin, ρout=ρout,
-                eigenvalues=eigenvalues, occupation=occupation, εF=εF, n_iter=n_iter, ψ=ψ,
-                diagonalization=nextstate.diagonalization, stage=:iterate)
+        info = merge(info, (energies=energies, ))
 
         # Apply mixing and pass it the full info as kwargs
-        ρnext = mix(mixing, basis, ρin, ρout; info...)
+        # TODO Mixing should take both density and spin density
```

Comment:
probably better to decide on an array API before doing the mixing thing

Reply:
Yes. We have to change the mixing interface as well. That's a bit quirky, too.